### PR TITLE
Changes to GMX tools

### PR DESCRIPTION
**Pull request by simonbray**
- [ ] <del>acpype tool currently failing on usegalaxy.eu since last version update</del> not something to fix on GitHub
- [x] split setup into separate pdb2gmx and editconf tools
- [x] make ion adding optional in solvate tool
- [x] options to return other files
   - [x] traj from EM tool
   - [x] edr files
   - [x] xvg files
- [ ] <del>consistent naming of restraints file: posres.itp or posre.itp</del> not an issue I think, or if so with GROMACS and not with our tools
- [ ] <del>allow upload of multiple config files?</del> don't recall why this was necessary. ndx files at least can now be uploaded as specified below
- [x] tool for creating ndx files?
- [x] allow upload of custom ndx file for sim tool
- [x] check behaviour when setting up proteins with multple peptide chains
- [x] add tool to extract energies from EDR files
- [x] add tool to create custom restraint files

### CODE REVIEW

**Comment by tsenapathi** [comment]
@simonbray is it possible to add a potion to plot the energies?

**Comment by simonbray** [comment]
@tsenapathi my idea was that the xvg file can be plotted using the existing Galaxy line plot visualization. You just need to remove the header lines to turn it into a simple tabular file, right?

**Comment by tsenapathi** [comment]
@simonbray true! but I think even a rough plot of the xvg file will be useful to visualize the data quickly or maybe add an option to plot or not plot?

**Comment by simonbray** [comment]
Maybe it would be nice to offer as an option. On the other hand I get a bit worried about creating a lot of options and making the tool too complex.

Thanks for the suggestion, I'll think about it. :)

File: tools/gromacs/energy.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+
+    <expand macro="requirements" />
+
+    <command detect_errors="exit_code"><![CDATA[
+
+        ln -s '$edr_input' ./edr_input.edr &&
+        echo $terms | tr , ' ' | gmx energy -f ./edr_input.edr -o ./energy.xvg
```
Comment (simonbray) [inline tools/gromacs/energy.xml]: @bgruening this works but I wonder if there is a nicer or more 'Galaxy-approved' way to do it.

The select parameter with `multiple="true"` gives me a comma-separated list and I want to turn it into a space-separated list and pipe it into `gmx energy`.

Comment (bgruening) [reply tools/gromacs/energy.xml]: Something like this: `' '.join($terms.split(','))`

**Comment by simonbray** [comment]
Ready to review+merge. :)

File: tools/gromacs/setup.xml [hunk 2]
```diff
@@ -9,13 +9,11 @@
     <command detect_errors="exit_code"><![CDATA[
 
         ln -s '$pdb_input' ./pdb_input.pdb &&
```
Comment (bgruening) [inline tools/gromacs/setup.xml]: to we need to bump the tool version here?

File: tools/gromacs/energy.xml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        ln -s '$edr_input' ./edr_input.edr &&
+        echo #echo str($terms).replace(',', ' ')# | gmx energy -f ./edr_input.edr -o ./energy.xvg
+        ## echo #echo ' '.join(str($terms).split(','))# | gmx energy -f ./edr_input.edr -o ./energy.xvg
+
+    ]]></command>
+
+    <inputs>
+        <param argument="edr_input" type="data" format="edr" label="EDR file" help="GROMACS energy binary file"/>
+        <param argument="terms" type='select' label="Terms to calculate" multiple="true" help="Select terms for which values should be calculated">
+            <option value="0">Angle</option>
```
Comment (bgruening) [inline tools/gromacs/energy.xml]: is angle a good default choice ... maybe use selected="True" to be more explicit

File: tools/gromacs/editconf.xml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    <tests>
+        <test>
+            <param name="input_file" value="processed.gro" ftype="gro"/>
+            <param name="output_format" value="gro" />
+            <param name="config" value="true" />
+            <param name="dim" value="1.0" />
+            <param name="type" value="cubic" />
+            <output name="output" file="newbox.gro"/>
+        </test>
+        <!-- <test>
```
Comment (bgruening) [inline tools/gromacs/editconf.xml]: ping? remove or readd?

File: tools/gromacs/editconf.xml [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                    <option value="cubic">Rectangular box with all sides equal</option>
+                    <option value="triclinic">Triclinic</option>
+                    <option value="dodecahedron">represents a rhombic dodecahedron</option>
+                    <option value="octahedron">truncated octahedron</option>
+                </param>
+            </when>
+            <when value="false"/>
+        </conditional>
+
+        <param name="capture_log" type="boolean" value="false" label="Generate detailed log" help="Generate detailed log information that can be summarized with ParseLog."/>
```
Comment (bgruening) [inline tools/gromacs/editconf.xml]: should this be a macro?

File: tools/gromacs/editconf.xml [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        gmx editconf
+            -f ./input.$input_file.ext
+            -o ./output.${output_format}
+            #if $box.config == "true":
+                -d $box.dim 
+                -bt $box.type
+            #end if
+            &>> verbose.txt &&
+
+        cp ./output.${output_format} $output
```
Comment (bgruening) [inline tools/gromacs/editconf.xml]: single quote output

File: tools/gromacs/makendx.xml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        <param argument="sel" type="text" label="Selection command" help="Selection for the new index group, e.g. 'r34' to select residue 34. Use  the operators '|' (or) and &amp; (and) to combine predefined groups.">
+            <sanitizer>
+                <valid>
+                    <add value="|" />
+                    <add value="&amp;" />
+                    <add value="!" />
+                </valid>
+            </sanitizer>
+        </param>
+        <param name="capture_log" type="boolean" value="false" label="Generate detailed log" help="Generate detailed log information that can be summarized with ParseLog."/>
```
Comment (bgruening) [inline tools/gromacs/makendx.xml]: macro?

File: tools/gromacs/makendx.xml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        <import>macros.xml</import>
+    </macros>
+
+    <expand macro="requirements" />
+
+    <command detect_errors="exit_code"><![CDATA[
+
+        ln -s '$input_file' ./input.$input_file.ext &&
+        (echo '$sel'; echo 'q') | gmx make_ndx -f ./input.$input_file.ext -o ./ndx.ndx &>> verbose.txt &&
+        cp ./ndx.ndx $ndx
```
Comment (bgruening) [inline tools/gromacs/makendx.xml]: single quote

File: tools/gromacs/restraints.xml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        <param argument="gro_input" type="data" format='gro' label="GRO input file" help="Input structure"/>
+        <param argument="ndx_input" type="data" format='ndx' label="Index (NDX) file" help="Index file containing the atoms which should be restrained"/>
+        <param argument="fc" type="text" value="1000 1000 1000" label="Force constants" help="Vector (in Cartesian space) specifying force constants for the restraints (kJ/mol nm^2)">
+            <validator type="regex" message="Only numeric values and whitespace allowed">^[0-9 ]*$</validator>
+        </param>
+        <param argument="index" type="text" label="Index of group" help="Index of group for calculating position restraints - i.e. the group's position in the ndx file (using zero-based numbering, so the first group has index 0).">
+            <validator type="regex" message="Only numeric values allowed">^[0-9]*$</validator>
+        </param>
+
+        <param name="capture_log" type="boolean" value="false" label="Generate detailed log" help="Generate detailed log information that can be summarized with ParseLog."/>
```
Comment (bgruening) [inline tools/gromacs/restraints.xml]: macro

File: tools/gromacs/sim.xml [hunk 10]
```diff
@@ -119,67 +122,120 @@
...
+        <section name="inps" title="Inputs" expanded="false">
+        
+            <!-- CPT inp -->
+            <conditional name="cpt_inp">
+                <param name="cpt_bool" type="select" label="Use a checkpoint (CPT) file" help="CPT file from a previous MD run">
+                    <option value="yes">Continue simulation from a CPT file</option>
+                    <option value="no" selected="true">No CPT input</option>
+                </param>
+                <when value="yes">
+                    <param argument="cpt_in" type="data" format='cpt' label="Checkpoint file"/>
```
Comment (bgruening) [inline tools/gromacs/sim.xml]: @simonbray why can this be not simply optional="true" and remove the entire conditional? Conditional with just one single param inside are not useful imho.

File: tools/gromacs/sim.xml [hunk 11]
```diff
@@ -119,67 +122,120 @@
...
+
+
+            <!-- ITP inp -->
+            <conditional name="posres">
+                <param name="posres_bool" type="select" label="Apply position restraints" help="Used e.g. for equilibration of solvent around a protein" value="false">
+                    <option value="true">Apply position restraints</option>
+                    <option value="false" selected="true">No position restraints</option>
+                </param>
+                <when value="true">
+                    <param argument="itp_inp" type="data" format='itp' label="Position restraint (ITP) file"/>
```
Comment (bgruening) [inline tools/gromacs/sim.xml]: See above.

File: tools/gromacs/sim.xml [hunk 12]
```diff
@@ -119,67 +122,120 @@
...
+
+
+            <!-- NDX inp -->
+            <conditional name="ndx_inp">
+                <param name="ndx_bool" type="select" label="Use a custom index (ndx) file" help="Use an index file specifying custom atom groups.">
+                    <option value="true">Use custom index file</option>
+                    <option value="false" selected="true">Use default generated group</option>
+                </param>
+                <when value="true">
+                    <param argument="ndx_in" type="data" format='ndx' label="Index file"/>
```
Comment (bgruening) [inline tools/gromacs/sim.xml]: See above.

File: tools/gromacs/editconf.xml (hunk 5 above)
Comment (simonbray) [reply tools/gromacs/editconf.xml]: yes

File: tools/gromacs/sim.xml (hunk 10 above)
Comment (simonbray) [reply tools/gromacs/sim.xml]: the results of the boolean are also fed into the config file. I don't know if there is another way of doing this.

File: tools/gromacs/setup.xml (hunk 2 above)
Comment (simonbray) [reply tools/gromacs/setup.xml]: done

**Comment by simonbray** [comment]
@bgruening, can we also merge this? I did not fix the boolean/conditional issue because the boolean is also needed in the config file.

File: tools/gromacs/sim.xml (hunk 10 above)
Comment (bgruening) [reply tools/gromacs/sim.xml]: @simonbray can you not do something like that for the boolean?

```
#if $file:
   set boolean
#end if
```
In the config file. So only set the boolean when the file was selected?

Comment (simonbray) [reply tools/gromacs/sim.xml]: @bgruening, done

**Comment by simonbray** [comment]
@bgruening, @tsenapathi, @chrisbarnettster, could one of you please merge this?

**Comment by simonbray** [comment]
Thanks Björn!

**Pull request merged**